crates/typst-macros/src/elem.rs

```diff
@@ -17,17 +17,99 @@ struct Elem {
     docs: String,
     vis: syn::Visibility,
     ident: Ident,
+    enum_ident: Ident,
     capabilities: Vec<Ident>,
     fields: Vec<Field>,
 }
 
-/// Details about an element field.
+impl Elem {
+    fn without_capability(
+        &self,
+        name: &str,
+        closure: impl FnOnce() -> TokenStream,
+    ) -> Option<TokenStream> {
+        self.capabilities
+            .iter()
+            .all(|capability| capability != name)
+            .then(closure)
+    }
+
+    fn with_capability(
+        &self,
+        name: &str,
+        closure: impl FnOnce() -> TokenStream,
+    ) -> Option<TokenStream> {
+        self.capabilities
+            .iter()
+            .any(|capability| capability == name)
+            .then(closure)
+    }
+
+    /// All fields.
+    ///
+    /// This includes:
+    /// - Fields that are not external and therefore present in the struct.
+    fn all_fields(&self) -> impl Iterator<Item = &Field> + Clone {
+        self.fields.iter().filter(|field| !field.external)
+    }
+
+    /// Fields that are inherent to the element.
+    fn inherent_fields(&self) -> impl Iterator<Item = &Field> + Clone {
+        self.all_fields().filter(|field| field.inherent())
+    }
+
+    /// Fields that can be set with style rules.
+    ///
+    /// This includes:
+    /// - Fields that are not synthesized.
+    /// - Fields that are not inherent and therefore present at all times.
+    /// - Fields that are not internal or have a parser.
+    fn settable_fields(&self) -> impl Iterator<Item = &Field> + Clone {
+        self.all_fields().filter(|field| {
+            !field.synthesized
+                && field.settable()
+                && (!field.internal || field.parse.is_some())
+        })
+    }
+
+    /// Fields that are visible to the user.
+    ///
+    /// This includes:
+    /// - Fields that are not internal or have a parser.
+    fn visible_fields(&self) -> impl Iterator<Item = &Field> + Clone {
+        self.all_fields()
+            .filter(|field| !field.internal || field.parse.is_some())
```

Comment:
should internal fields with a parser really be shown? typically, they are internal for a reason.